plugin.rb

```diff
@@ -46,4 +46,63 @@ class Engine < ::Rails::Engine
       DiscourseLivestream.handle_topic_chat_channel_creation(topic)
     end
   end
+
+  register_modifier(:list_user_channels_modifier) do |f, user|
+    user_livestream_chat_channel_memberships =
+      Chat::UserChatChannelMembership
+        .joins(chat_channel: { livestream_topic_chat_channel: :topic })
+        .where(user: user)
+        .includes(chat_channel: { livestream_topic_chat_channel: :topic })
+
+    user_allowed_groups = SiteSetting.livestream_chat_allowed_groups.split("|").map(&:to_i)
+    user_group_ids = user.groups.ids
+    user_allowed_in_topic_chat_channels = (user_allowed_groups & user_group_ids).any?
+
+    if SiteSetting.calendar_enabled && SiteSetting.discourse_post_event_enabled
+      user_livestream_chat_channel_memberships.each do |membership|
+        topic_chat_channel = membership.chat_channel.livestream_topic_chat_channel
+        next unless topic_chat_channel
+
+        event_invitee =
+          topic_chat_channel.topic.posts.first&.event&.invitees&.find_by(user_id: user.id)
+        next unless event_invitee
+
+        invitee_status = event_invitee.status
+        is_going = invitee_status == DiscoursePostEvent::Invitee.statuses[:going]
+        if user_allowed_in_topic_chat_channels && is_going && !membership.following
+          Chat::ChannelMembershipManager.new(membership.chat_channel).follow(user)
+        elsif !user_allowed_in_topic_chat_channels && is_going && membership.following
+          Chat::ChannelMembershipManager.new(membership.chat_channel).unfollow(user)
+        end
+      end
+    end
+    Chat::UserChatChannelMembership.where(user: user)
+  end
+
+  register_modifier(:follow_modifier) do |f, channel, user, membership, object|
+    topic_chat_channel = DiscourseLivestream::TopicChatChannel.find_by(chat_channel_id: channel.id)
+
+    user_allowed_groups = SiteSetting.livestream_chat_allowed_groups.split("|").map(&:to_i)
+    user_group_ids = user.groups.ids
+    user_allowed_in_topic_chat_channels = (user_allowed_groups & user_group_ids).any?
+    if topic_chat_channel && !user_allowed_in_topic_chat_channels
+      ActiveRecord::Base.transaction do
```

Comment:
You could move the transaction up to use a single transaction for both statements. Also, I’d avoid nesting altogether. Maybe something like this would work:

```ruby
ActiveRecord::Base.transaction do
  if topic_chat_channel && !user_allowed_in_topic_chat_channels
    return membership unless membership.following

    membership.update!(following: false)
    object.recalculate_user_count
    return membership
  end

  if membership.new_record?
    membership.save!
    object.recalculate_user_count
    return membership
  end

  if !membership.following
    membership.update!(following: true)
    object.recalculate_user_count
    return membership
  end
end
```